crates/core/src/octree/build/stages/allocate_nodes.rs

```diff
@@ -57,3 +63,157 @@ impl AllocateNodesPass {
         self.shader.wait();
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::path::PathBuf;
+    use std::env;
+
+    struct TestCase {
+        pub input: AllocateNodesInput,
+        pub expected_output: Vec<u32>,
+        pub description: String,
+    }
+
+    #[test]
+    fn allocate_nodes_works() {
+        let (_glfw, _window) = test_utils::init_opengl_context();
+
+        // To go from the crate root to the workspace root
+        // This is stateful, so not a good testing practice, doing it on first test ran for now
+        let mut path = PathBuf::from(env::current_dir().unwrap());
+        path.pop();
+        path.pop();
+        env::set_current_dir(path).unwrap();
+
+        let voxel_dimension_exponent = 4;
+
+        unsafe {
+            let test_data = test_cases();
+
+            // Initialize everything
+            Config::initialize_test_sensitive(Config::new(voxel_dimension_exponent), true);
+            let allocate_nodes_pass = AllocateNodesPass::init();
+            let mut allocations: Vec<u32> = vec!();
+
+            for TestCase { input, expected_output, description } in test_data.iter() {
+                // Run the shader
+                allocate_nodes_pass.run(input.clone());
+
+                // Verify output
+                let output = input.node_pool.data();
+                println!("Running test: {description}");
+                // Both the for and the later assert are because the allocated node number is non
+                // deterministic, as it depends on each thread calling to the allocation counter
+                for (index, &item) in expected_output.iter().enumerate() {
+                    if item != u32::MAX {
+                        assert_eq!(output[index], item);
+                    } else {
+                        allocations.push(output[index].clone());
+                    }
+                }
+                let expected_allocations: Vec<u32> = (input.first_free_node as u32..input.first_free_node as u32 + allocations.len() as u32 ).collect();
+                allocations.sort();
+                assert_eq!(*allocations, *expected_allocations);
+                allocations.clear();
+                println!("Passed");
+            }
+        }
+    }
+
+    unsafe fn test_cases() -> std::vec::Vec<TestCase> {
```

Comment:
I'd love to test the case where we add border data to the mix
We should also look into what edge cases we can test, such as number of threads

Reply:
That is probably more in the side of integration tests right?

Reply:
Probably